src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph.cs

```diff
@@ -14,35 +14,288 @@ namespace Microsoft.CodeAnalysis
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> existingReferencesMap,
+            ProjectId projectId,
+            IReadOnlyList<ProjectId> referencedProjectIds)
+        {
+            if (existingReferencesMap.TryGetValue(projectId, out var existingReferences))
+            {
+                return existingReferencesMap.SetItem(projectId, existingReferences.Union(referencedProjectIds));
+            }
+            else
+            {
+                return existingReferencesMap.SetItem(projectId, referencedProjectIds.ToImmutableHashSet());
+            }
+        }
+
+        /// <summary>
+        /// Computes a new <see cref="_lazyReverseReferencesMap"/> for the addition of additional project references. Must be called on a non-null
+        /// map.
+        /// </summary>
+        private static ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> ComputeNewReverseReferencesMapForAdditionalProjectReferences(
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> existingReverseReferencesMap,
+            ProjectId projectId,
+            IReadOnlyList<ProjectId> referencedProjectIds)
+        {
+            var builder = existingReverseReferencesMap.ToBuilder();
+
+            foreach (var referencedProject in referencedProjectIds)
+            {
+                if (builder.TryGetValue(referencedProject, out var reverseReferences))
+                {
+                    builder[referencedProject] = reverseReferences.Add(projectId);
+                }
+                else
+                {
+                    builder[referencedProject] = ImmutableHashSet.Create(projectId);
+                }
+            }
+
+            return builder.ToImmutable();
+        }
+
+        /// <summary>
+        /// Computes a new <see cref="_transitiveReferencesMap"/> for the addition of additional project references. 
+        /// </summary>
+        private static ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> ComputeNewTransitiveReferencesMapForAdditionalProjectReferences(
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> existingTransitiveReferencesMap,
+            ProjectId projectId,
+            IReadOnlyList<ProjectId> referencedProjectIds)
+        {
+            // To update our forward transitive map, we need to add referencedProjectIds (and their transitive dependencies) to the transitive references
+            // of projects. First, let's just compute the new set of transitive references. It's possible while doing so we'll discover that we don't
+            // know the transitive project references for one of our new references. In that case, we'll use null as a sentinel to mean "we don't know" and
+            // we propogate the not-knowingness. But let's not worry about that yet. First, let's just get the new transitive reference set.
+            var newTransitiveReferences = new HashSet<ProjectId>(referencedProjectIds);
```

Comment:
can you wrap comments around 120 cols?

Reply:
Ok.  dumb question before i go review all this... why is updating transitive references not simple (given that we already have an *existing* valid direct-reverse-reference map?

So i am adding references R1 to RN to Project P1.  Ok.  So i first add it to the 'trans references' set for P1.  I then use my reverse dep map to see who references P1, i get Pq, Pr, Ps.  I go add these references to those three projects.  I then recurse on those.   And i basically back propagate the references added to P1, backwards updating anything as necessary.

Why is that not sufficient?

Reply:
I've answered this in your other conversation with the pseudocode.